drivers/spi/spi_ll_stm32.c

```diff
@@ -716,6 +717,7 @@ static int transceive_dma(struct device *dev,
 
 		/* once SPI is no more busy, wait for DMA transfer end */
 		while (spi_stm32_dma_transfer_ongoing(data) == 1) {
+			k_sem_take(&data->dma_status_sem, K_MSEC(100));
```

Comment:
A semaphore is definitely better than the busy loop, as it allows the CPU to execute other tasks or sleep. A busy loop kind of defeat the purpose of DMA...

Now I have two points:
- The underlying bug is that the `transfer_complete` bool in struct stream is not declared volatile, so the compiler optimizes the loop as a if test and an endless loop. I am afraid that even after your change, the bug is still there. Now that there is a semaphore, I believe that the `transfer_complete` booleans  are redundant an not needed anymore.
- You used a timeout in `k_sem_take`. I would suggest either to use `K_FOREVER` or to properly handle the timeout by checking the value returned this function

Reply:
If the DMA flags are just flags and not volatile (or atomic, or some sort of MB) that seems broken by design :-/ I didn't run into that problem yet, will take a look if it also can cause me problems. 

The timeout was thought as preparation for a real timeout so it will be possible to recover from when "something" goes badly goes wrong (better one broken transfer than a system that hangs for ever). For now it shouldn't matter that the k_sem_take wakes up every 100ms, it will check if the dma is still ongoing and go back to sleep.  But if it is a show-stopper I can make a K_FOREVER out of it.

Reply:
> * The underlying bug is that the `transfer_complete` bool in struct stream is not declared volatile, so the compiler optimizes the loop as a if test and an endless loop. I am afraid that even after your change, the bug is still there. Now that there is a semaphore, I believe that the `transfer_complete` booleans  are redundant an not needed anymore.

I have one semaphore for signaling RX and TX state (the two flags still say if RX and TX are both ready), maybe removing those flags and using 2 semaphores to signal TX and RX are done. One could first take TX and than take RX. 

PS: the k_sem_take in the loop might fix your optimizing problem because the k_sem_take code surely has some kind of memory barrier code in it so the compiler can't optimize the loop away anymore.

Reply:
> > * The underlying bug is that the `transfer_complete` bool in struct stream is not declared volatile, so the compiler optimizes the loop as a if test and an endless loop. I am afraid that even after your change, the bug is still there. Now that there is a semaphore, I believe that the `transfer_complete` booleans  are redundant an not needed anymore.
> 
> I have one semaphore for signaling RX and TX state (the two flags still say if RX and TX are both ready), maybe removing those flags and using 2 semaphores to signal TX and RX are done. One could first take TX and than take RX.

Yes that indeed looks like the way to go, right now the semaphores and the booleans provide a redundant information.

> PS: the k_sem_take in the loop might fix your optimizing problem because the k_sem_take code surely has some kind of memory barrier code in it so the compiler can't optimize the loop away anymore.

Yes that's a good point. Actually given that k_sem_take is defined in a different file and LTO is not enabled, the compiler can't do any optimization anymore.

Reply:
Just realized that one sem should be enough it just has to be given twice (for TX and RX) and taken twice, since the code is only interested in knowning that both RX and TX are done. But on the other hand that does mean it must be absolutly certain that both RX and TX DMA complete signals come only 1 time, so maybe a TX and RX sem is more robust.